src/librustc_driver/lib.rs

```diff
@@ -383,17 +383,12 @@ pub fn run_compiler(
             mem::drop(compiler.expansion()?.take());
         }
 
-        compiler.ongoing_codegen()?;
-
-        // Drop GlobalCtxt after starting codegen to free memory
-        mem::drop(compiler.global_ctxt()?.take());
+        compiler.codegen_and_link()?;
```

Comment:
having reviewed @Zoxc's PRs it seems to me like only this commit is actually problematic for the querification PRs (like those PRs would have to revert this commit to be doable). As far as I can tell (both from the code and your graphs) it's not that much additional complexity to keep around either.

What do you think?

Reply:
Thank you for the feedback. Are you referring to #59404? If so:
- If there's a good reason to keep the `ongoing_codegen`/`link` split then I am fine to remove this commit.
- Is there a good reason? Does #59404 and its predecessors use the existing passes because they are there, rather than because they are necessary?

Reply:
> * If there's a good reason to keep the `ongoing_codegen`/`link` split then I am fine to remove this commit.

The main reason to keep this separate because `ongoing_codegen` needs the `GlobalCtxt` to be around, while we want to free `GlobalCtxt` before linking. When we remove queries from `rustc_interface` this mean this cannot be a combined operation.

Another reason is that it's reasonable for front-ends to do only code generation, but no linking, although that is not yet possible.

Reply:
I don't understand the point about `GlobalCtxt`. My patch preserves the "free `GlobalCtxt` before linking" behaviour.

As for the second point, I would argue that YAGNI applies.

Reply:
The problem isn't that your PR doesn't keep the early drop, it's the way the query PR works, where the `GlobalCtxt` lives as long as a closure and is automatically dropped at the end of it. So you can't call the combined function inside the closure and you can't call it outside the closure, thus needing to split it up again.